File: C#/118. Pascal's Triangle.cs
/*
 * Link: https://leetcode.com/problems/pascals-triangle/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    public IList<IList<int>> Generate(int numRows) {
        if (numRows == 0) return new List<IList<int>>();
        
        var res = new List<IList<int>>();
        
        res.Add(new List<int>() {1});
        if (numRows == 1) return res;
        
        for (var r = 2; r <= numRows; r++) {
            var preRow = res[r - 2];
            
            var currentRow = new List<int>() {};
            
            for (var col = 0; col < r; col++) {
                int top = (col == r - 1) ? 0 : preRow[col];
                int topLeft = (col == 0) ? 0 : preRow[col - 1];
                
                currentRow.Add(top + topLeft);
            }
            
            res.Add(currentRow);
        }
        
        return res;
    }
}

File: Python/118. Pascal's Triangle.py
class Solution(object):
    def generate(self, numRows):
        """
        :type numRows: int
        :rtype: List[List[int]]
        """
        res = [[1]]
        cur, prev = [1], []
        if numRows == 1:
            return res
        for r in range(2, numRows + 1):
            prev = cur
            cur = []
            for c in range(r):
                leftVal = 0 if c == 0 else prev[c - 1]
                curVal = prev[c] if c < len(prev) else 0
                cur.append(leftVal + curVal)
            res.append(cur)
        return res